src/org/intellij/erlang/completion/ErlangAppCompletionContributor.java

```diff
@@ -0,0 +1,100 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.intellij.erlang.completion;
+
+import com.intellij.codeInsight.completion.*;
+import com.intellij.codeInsight.lookup.LookupElementBuilder;
+import com.intellij.patterns.PatternCondition;
+import com.intellij.patterns.PsiElementPattern;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.util.PsiTreeUtil;
+import com.intellij.util.ProcessingContext;
+import com.intellij.util.containers.ContainerUtil;
+import org.intellij.erlang.ErlangFileType;
+import org.intellij.erlang.psi.*;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+
+import static com.intellij.patterns.PlatformPatterns.psiElement;
+import static com.intellij.patterns.PlatformPatterns.virtualFile;
+import static com.intellij.patterns.StandardPatterns.instanceOf;
+
+public class ErlangAppCompletionContributor extends CompletionContributor {
+  // source: "FILE SYNTAX" section of http://www.erlang.org/doc/man/app.html
+  static final List<String> KEYWORDS =
+    ContainerUtil.immutableList("description", "id", "vsn", "modules", "maxP", "maxT", "registered", "applications",
+                                "included_applications", "env", "mod", "start_phases", "runtime_dependencies");
+  private static final String APPLICATION_KEYWORD = "application";
+
+  @Override
+  public void beforeCompletion(@NotNull CompletionInitializationContext context) {
+    PsiFile file = context.getFile();
+    if (file.getFileType() != ErlangFileType.APP) return;
+    int startOffset = context.getStartOffset();
+    PsiElement elementAt = file.findElementAt(startOffset);
+    PsiElement parent = elementAt != null ? elementAt.getParent() : null;
+    if (parent instanceof ErlangTupleExpression) {
+      context.setDummyIdentifier(APPLICATION_KEYWORD);
+    }
+  }
+
+  public ErlangAppCompletionContributor() {
+    PsiElementPattern.Capture<PsiElement> psiElementInAppFile =
+      psiElement().inFile(instanceOf(ErlangFile.class)).inVirtualFile(virtualFile().ofType(ErlangFileType.APP));
+
+    //noinspection unchecked
+    extend(CompletionType.BASIC,
+           psiElementInAppFile.withParents(ErlangAtom.class, ErlangQAtom.class, ErlangConfigExpression.class,
+                                           ErlangTupleExpression.class, ErlangFile.class).with(positionInTuple(0)),
+           getProvider(ContainerUtil.list(APPLICATION_KEYWORD)));
+    //noinspection unchecked
+    extend(CompletionType.BASIC,
+           psiElementInAppFile.withParents(ErlangAtom.class, ErlangQAtom.class, ErlangConfigExpression.class,
+                                           ErlangTupleExpression.class, ErlangListExpression.class,
+                                           ErlangTupleExpression.class, ErlangFile.class).with(positionInTuple(0)),
```

Comment:
Also, why do you keep this ```with(positionInTuple(0))``` in both calls - why not raise move it to a variable you already have?